core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -949,9 +953,65 @@ public void onSuccess() {
                                         }
 
                                         @Override
-                                        public void onFailure(Throwable t) {
-                                            // TODO: handle catastrophic non-channel failures
-                                            onReplicaFailure(nodeId, exp);
+                                        public void onFailure(Throwable shardFailedError) {
+                                            logger.error("[{}] catastrophic error while failing replica shard [{}] for [{}]", shardFailedError, shardId, shard, exp);
+                                            if (shardFailedError instanceof ShardStateAction.NoLongerPrimaryShardException) {
+                                                // we are no longer the primary, fail ourselves and start over
+                                                ShardRouting primaryShard = indexShardReference.routingEntry();
+                                                String message = String.format(Locale.ROOT, "primary shard [%s] was demoted while failing replica shard [%s] for [%s]", primaryShard, shard, exp);
+                                                shardStateAction.shardFailed(primaryShard, primaryShard, message, shardFailedError, new ShardStateAction.Listener() {
```

Comment:
I think this should be done via IndexShard#failShard (which can be exposed via indexShardReference ). Will be cleaner and faster (it's local fail first, then notify the master)